File: crypto/src/main/java/net/corda/v5/crypto/DigitalSignature.java
package net.corda.v5.crypto;

import net.corda.v5.base.annotations.CordaSerializable;
import net.corda.v5.base.annotations.DoNotImplement;
import org.jetbrains.annotations.NotNull;

/**
 * A wrapper around a digital signature.
 */
@DoNotImplement
@CordaSerializable
public interface DigitalSignature {

    @NotNull
    byte[] getBytes();

    /**
     * A digital signature that identifies who is the owner of the signing key used to create this signature.
     */
    @DoNotImplement
    interface WithKeyId extends DigitalSignature {

        /**
         * Gets the key ID of the public key (public key hash) whose private key pair was used to sign the data. If the
         * original key passed in to the sign operation is a {@link CompositeKey} then the key ID, is the ID of the
         * composite key leaf used to sign.
         */
        @NotNull
        SecureHash getBy();
    }
}

File: crypto/src/main/kotlin/net/corda/v5/crypto/DigitalSignature.kt
package net.corda.v5.crypto

import net.corda.v5.base.annotations.CordaSerializable
import net.corda.v5.base.types.OpaqueBytes
import java.security.PublicKey

/**
 * A wrapper around a digital signature.
 *
 * @property bytes The signature.
 * @param bytes The signature.
 */
@CordaSerializable
open class DigitalSignature(
    bytes: ByteArray
) : OpaqueBytes(bytes) {
    /**
     * A digital signature that identifies who the public key is owned by.
     *
     * @param by The public key of the corresponding private key used to sign the data (as if an instance
     * of the [CompositeKey] is passed to the sign operation it may contain keys which are not actually owned by
     * the member).
     * @param bytes The signature.
     * @param context The context which was passed to the signing operation, note that this context is not signed over.
     */
    open class WithKey(
        /**
         * Public key which corresponding private key was used to sign the data (as if an instance
         * of the [CompositeKey] is passed to the sign operation it may contain keys which are not actually owned by
         * the member).
         */
        val by: PublicKey,
        bytes: ByteArray,
        /**
         * The context which was passed to the signing operation, note that this context is not signed over.
         */
        val context: Map<String, String>
    ) : DigitalSignature(bytes)
}
